extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduRangeRead.java

```diff
@@ -1,189 +1,251 @@
-                    upperPred).addPredicate(partitionPred).build();
-            KuduUtils.executeQuery(scanner, results);
-          }
+    final List<AsyncKuduScanner> scanners = new ArrayList<>();
+
+    final KuduPredicate adapterIdPred =
+        KuduPredicate.newInListPredicate(
+            schema.getColumn(KuduField.GW_ADAPTER_ID_KEY.getFieldName()),
+            Arrays.asList(ArrayUtils.toObject(adapterIds)));
+    if (ranges != null && !ranges.isEmpty()) {
+      for (final SinglePartitionQueryRanges r : ranges) {
+        byte[] partitionKey = r.getPartitionKey();
+        if (partitionKey == null) {
+          partitionKey = KuduUtils.EMPTY_KEY;
         }
-      } else if (dataIds != null) {
-        final String partitionIdField =
-            isDataIndex ? KuduDataIndexField.GW_PARTITION_ID_KEY.getFieldName()
-                : KuduField.GW_PARTITION_ID_KEY.getFieldName();
-        for (final byte[] dataId : dataIds) {
-          KuduPredicate partitionPred =
+        final KuduPredicate partitionPred =
+            KuduPredicate.newComparisonPredicate(
+                schema.getColumn(KuduField.GW_PARTITION_ID_KEY.getFieldName()),
+                ComparisonOp.EQUAL,
+                partitionKey);
+        for (final ByteArrayRange range : r.getSortKeyRanges()) {
+          final byte[] start = range.getStart() != null ? range.getStart() : new byte[0];
+          final byte[] end =
+              range.getEnd() != null ? range.getEndAsNextPrefix()
+                  : new byte[] {
+                      (byte) 0xFF,
+                      (byte) 0xFF,
+                      (byte) 0xFF,
+                      (byte) 0xFF,
+                      (byte) 0xFF,
+                      (byte) 0xFF,
+                      (byte) 0xFF};
+          final KuduPredicate lowerPred =
+              KuduPredicate.newComparisonPredicate(
+                  schema.getColumn(KuduField.GW_SORT_KEY.getFieldName()),
+                  ComparisonOp.GREATER_EQUAL,
+                  start);
+          final KuduPredicate upperPred =
               KuduPredicate.newComparisonPredicate(
-                  schema.getColumn(partitionIdField),
-                  ComparisonOp.EQUAL,
-                  dataId);
-          KuduScannerBuilder scannerBuilder = operations.getScannerBuilder(table);
-          KuduScanner scanner =
-              scannerBuilder.addPredicate(adapterIdPred).addPredicate(partitionPred).build();
-          KuduUtils.executeQuery(scanner, results);
+                  schema.getColumn(KuduField.GW_SORT_KEY.getFieldName()),
+                  ComparisonOp.LESS,
+                  end);
+
+          final AsyncKuduScannerBuilder scannerBuilder = operations.getAsyncScannerBuilder(table);
+          final AsyncKuduScanner scanner =
+              scannerBuilder.addPredicate(lowerPred).addPredicate(upperPred).addPredicate(
+                  partitionPred).addPredicate(adapterIdPred).build();
```

Comment:
Can you reorder this so that `adapterIdPred` is added first? Then `partitionPred`, and then `lowerPred`/`upperPred`? It's so that the order that you define the predicates above is the same as the order that you add them